### ISSUE TITLE
The gradient backpropogation during in-batch negative training

### ISSUE THREAD

**Issue by SparkJiao** [facebookresearch/DPR#109]
Thanks for you contribution to the community!

I have a question about the backpropogation during in-batch negative training.

```
    distributed_world_size = args.distributed_world_size or 1
    if distributed_world_size > 1:
        q_vector_to_send = (
            torch.empty_like(local_q_vector).cpu().copy_(local_q_vector).detach_()
        )
        ctx_vector_to_send = (
            torch.empty_like(local_ctx_vectors).cpu().copy_(local_ctx_vectors).detach_()
        )

        global_question_ctx_vectors = all_gather_list(
            [
                q_vector_to_send,
                ctx_vector_to_send,
                local_positive_idxs,
                local_hard_negatives_idxs,
            ],
            max_size=args.global_loss_buf_sz,
        )
```

It seems that under DDP, the negative samples located at other GPU are removed from the computation graph so that the gradient will not be backwarded through the samples located on different GPU, right?

Thanks for your help!

**Comment by vlad-karpukhin** [comment]
Hello,
The vectors calculated on other gpus (for other ctx & questions) don't need to carry their computation graph to other gpus. In order to calculate the loss function, you just need detached vectors. Each gpu has its own computation graph calculated on its specific questions & passages. Once the loss function is computed, the backpropagation calculates correct gradients  specific for every gpu. Then the gradients are synced (aggregated) across all gpus.

**Comment by SparkJiao** [comment]
Thanks for your reply!

In my view, the gradients are backpropagated along the vectors of questions and contexts located on the specific gpu, and the vectors received from other devices are viewed as constant vectors and the gradients will not be backpropagated through them.
Finally, the gradients are averaged over all gpus.

Is my understanding correct ?

Thank you very much.

**Comment by vlad-karpukhin** [comment]
"the gradients are backpropagated" - the gradients are computed when backpropagation happens.
Since every gpus has its own questions and passages encoded in the forward pass, gradients will be computed using those 'local' input data AND global loss value which is computed using all the vectors from other gpus. You don't need other question/passages computation graph to compute loss and 'local' gradients.
BUT every question&passage will be accounted for in the 'local' gradients.

The current schema can be decoupled even more. Please have a look at this paper, for example:
https://arxiv.org/pdf/2012.12624.pdf

**Comment by SparkJiao** [comment]
Great, I've got it.
Thanks for your quick and clear reply :)

I'll close the issue.
